File: android/app/src/test/java/com/moviemate/app/data/recommendation/MatchEngineTest.kt
package com.moviemate.app.data.recommendation

import java.time.Instant
import kotlin.math.abs
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Ported from web/match-engine.test.js, itself ported from
 * functions/test/{tasteProfile,scoring,streak}.test.ts — the same logic,
 * the same edge cases, now pinned for the Kotlin copy. The two documented
 * deviations carry over unchanged: no country signal, and advanceStreak has
 * no timezone parameter (always UTC calendar days — see the last streak
 * test, which pins that deviation explicitly).
 */
class MatchEngineTest {

    private fun rated(filmId: String, score: Double, genres: List<String> = listOf("Drama"), releaseYear: Int = 2021) =
        RatedFilm(filmId = filmId, genres = genres, releaseYear = releaseYear, score = score)

    private fun candidate(
        filmId: String,
        genres: List<String> = listOf("Sci-Fi"),
        releaseYear: Int = 2021,
        tmdbRating: Double = 7.0,
    ) = ScorableFilm(filmId = filmId, genres = genres, releaseYear = releaseYear, tmdbRating = tmdbRating)

    // ---------- decadeKey ----------

    @Test
    fun `decadeKey buckets modern years by decade`() {
        assertEquals("2020s", decadeKey(2024))
        assertEquals("2020s", decadeKey(2020))
        assertEquals("2010s", decadeKey(2019))
        assertEquals("2000s", decadeKey(2000))
    }

    @Test
    fun `decadeKey collapses everything before 2000 into one bucket`() {
        assertEquals("pre-2000", decadeKey(1999))
        assertEquals("pre-2000", decadeKey(1954))
    }

    // ---------- buildTasteProfile ----------

    @Test
    fun `buildTasteProfile averages scores per genre across films`() {
        val profile = buildTasteProfile(
            listOf(
                rated("1", 80.0, genres = listOf("Sci-Fi")),
                rated("2", 60.0, genres = listOf("Sci-Fi")),
                rated("3", 20.0, genres = listOf("Comedy")),
            ),
        )

        assertEquals(70.0, profile.genreAffinity["Sci-Fi"]!!, 1e-9)
        assertEquals(20.0, profile.genreAffinity["Comedy"]!!, 1e-9)
        assertEquals(3, profile.sampleSize)
    }

    @Test
    fun `buildTasteProfile credits a multi-genre film to every one of its genres`() {
        val profile = buildTasteProfile(listOf(rated("1", 90.0, genres = listOf("Sci-Fi", "Drama"))))

        assertEquals(90.0, profile.genreAffinity["Sci-Fi"]!!, 1e-9)
        assertEquals(90.0, profile.genreAffinity["Drama"]!!, 1e-9)
    }

    @Test
    fun `buildTasteProfile builds era affinity alongside genre`() {
        val profile = buildTasteProfile(
            listOf(
                rated("1", 85.0, releaseYear = 2022),
                rated("2", 45.0, releaseYear = 1994),
            ),
        )

        assertEquals(85.0, profile.eraAffinity["2020s"]!!, 1e-9)
        assertEquals(45.0, profile.eraAffinity["pre-2000"]!!, 1e-9)
    }

    @Test
    fun `buildTasteProfile returns an empty profile for a user who has rated nothing`() {
        val profile = buildTasteProfile(emptyList())
        assertTrue(profile.genreAffinity.isEmpty())
        assertEquals(0, profile.sampleSize)
    }

    // ---------- predictScore ----------

    @Test
    fun `predictScore applies the configured genre-era signal weights`() {
        val profile = buildTasteProfile(listOf(rated("1", 100.0, genres = listOf("Sci-Fi"), releaseYear = 2021)))
        assertEquals(100.0, predictScore(profile, candidate("c1")), 1e-6)
    }

    @Test
    fun `predictScore weights genre more heavily than era`() {
        val genreOnly = buildTasteProfile(listOf(rated("1", 100.0, genres = listOf("Sci-Fi"), releaseYear = 1980)))
        val eraOnly = buildTasteProfile(listOf(rated("1", 100.0, genres = listOf("Horror"), releaseYear = 2021)))
        val target = candidate("c1")

        assertTrue(predictScore(genreOnly, target) > predictScore(eraOnly, target))
    }

    @Test
    fun `predictScore falls back to the neutral midpoint for unseen dimensions`() {
        val profile = buildTasteProfile(emptyList())
        assertEquals(AlgorithmConfig.NEUTRAL_AFFINITY, predictScore(profile, candidate("c1")), 1e-6)
    }

    @Test
    fun `predictScore stays inside 0-100 for the extremes of the Taste Dial`() {
        val hated = buildTasteProfile(listOf(rated("1", 0.0, genres = listOf("Sci-Fi"), releaseYear = 2021)))
        val loved = buildTasteProfile(listOf(rated("1", 100.0, genres = listOf("Sci-Fi"), releaseYear = 2021)))
        val target = candidate("c1")

        assertTrue(predictScore(hated, target) >= 0.0)
        assertTrue(predictScore(loved, target) <= 100.0)
    }

    // ---------- scoreCandidate ----------

    @Test
    fun `scoreCandidate penalises divergence - a shared 70-70 beats a lopsided 95-45`() {
        val sciFiFan = buildTasteProfile(
            listOf(
                rated("a1", 95.0, genres = listOf("Sci-Fi")),
                rated("a2", 95.0, genres = listOf("Sci-Fi"), releaseYear = 2022),
            ),
        )
        val sciFiSkeptic = buildTasteProfile(
            listOf(
                rated("b1", 45.0, genres = listOf("Sci-Fi")),
                rated("b2", 45.0, genres = listOf("Sci-Fi"), releaseYear = 2022),
            ),
        )
        val bothLukewarm = buildTasteProfile(
            listOf(
                rated("c1", 70.0, genres = listOf("Sci-Fi")),
                rated("c2", 70.0, genres = listOf("Sci-Fi"), releaseYear = 2022),
            ),
        )

        val target = candidate("x")
        val lopsided = scoreCandidate(sciFiFan, sciFiSkeptic, target)
        val shared = scoreCandidate(bothLukewarm, bothLukewarm, target)

        assertEquals(
            shared.breakdown.predictedA + shared.breakdown.predictedB,
            lopsided.breakdown.predictedA + lopsided.breakdown.predictedB,
            1e-6,
        )
        assertTrue(shared.finalScore > lopsided.finalScore)
    }

    @Test
    fun `scoreCandidate applies the divergence penalty at the configured rate`() {
        val a = buildTasteProfile(listOf(rated("1", 100.0, genres = listOf("Sci-Fi"))))
        val b = buildTasteProfile(listOf(rated("1", 0.0, genres = listOf("Sci-Fi"))))
        val result = scoreCandidate(a, b, candidate("x"))

        val breakdown = result.breakdown
        assertEquals(abs(breakdown.predictedA - breakdown.predictedB), breakdown.divergence, 1e-6)
        assertEquals(
            (breakdown.predictedA + breakdown.predictedB) / 2 - breakdown.divergence * AlgorithmConfig.DIVERGENCE_PENALTY,
            breakdown.tasteScore,
            1e-6,
        )
    }

    @Test
    fun `scoreCandidate scales the TMDB quality bonus onto the same 0-100 axis as taste`() {
        val profile = buildTasteProfile(listOf(rated("1", 70.0, genres = listOf("Sci-Fi"))))
        val perfect = scoreCandidate(profile, profile, candidate("c1", tmdbRating = 10.0))
        val awful = scoreCandidate(profile, profile, candidate("c2", tmdbRating = 0.0))

        assertEquals(100.0, perfect.breakdown.qualityBonus, 1e-9)
        assertEquals(0.0, awful.breakdown.qualityBonus, 1e-9)
        assertEquals(100.0 * AlgorithmConfig.QUALITY_WEIGHT, perfect.finalScore - awful.finalScore, 1e-6)
    }

    @Test
    fun `scoreCandidate lets shared taste outrank general acclaim, as the low quality weight intends`() {
        val loversOfHorror = buildTasteProfile(
            listOf(
                rated("1", 95.0, genres = listOf("Horror")),
                rated("2", 95.0, genres = listOf("Horror"), releaseYear = 2022),
            ),
        )
        val belovedButWrongGenre = candidate("acclaimed", genres = listOf("Musical"), tmdbRating = 9.5)
        val ourKindOfFilm = candidate("ours", genres = listOf("Horror"), tmdbRating = 6.0)

        val acclaimed = scoreCandidate(loversOfHorror, loversOfHorror, belovedButWrongGenre)
        val ours = scoreCandidate(loversOfHorror, loversOfHorror, ourKindOfFilm)

        assertTrue(ours.finalScore > acclaimed.finalScore)
    }

    @Test
    fun `scoreCandidate never returns a score outside 0-100`() {
        val hater = buildTasteProfile(listOf(rated("1", 0.0, genres = listOf("Sci-Fi"))))
        val lover = buildTasteProfile(listOf(rated("1", 100.0, genres = listOf("Sci-Fi"))))

        for ((a, b) in listOf(hater to hater, lover to lover, hater to lover)) {
            for (tmdbRating in listOf(0.0, 5.0, 10.0)) {
                val finalScore = scoreCandidate(a, b, candidate("x", tmdbRating = tmdbRating)).finalScore
                assertTrue(finalScore >= 0.0)
                assertTrue(finalScore <= 100.0)
            }
        }
    }

    // ---------- buildReason ----------

    @Test
    fun `buildReason names the shared genre when affinity clears the strong threshold`() {
        val fans = buildTasteProfile(
            listOf(
                rated("1", 90.0, genres = listOf("Comedy")),
                rated("2", 90.0, genres = listOf("Comedy"), releaseYear = 2022),
            ),
        )
        val reason = buildReason(fans, fans, candidate("x", genres = listOf("Comedy")))
        assertEquals("You both love Comedy", reason)
    }

    @Test
    fun `buildReason falls back to era when no genre clears the threshold`() {
        val fans = buildTasteProfile(
            listOf(
                rated("1", 90.0, genres = listOf("Drama"), releaseYear = 1985),
                rated("2", 90.0, genres = listOf("Drama"), releaseYear = 1988),
            ),
        )
        val reason = buildReason(fans, fans, candidate("x", genres = listOf("Western"), releaseYear = 1987))
        assertEquals("You both gravitate toward pre-2000s films", reason)
    }

    @Test
    fun `buildReason gives a generic reason when nothing clears either threshold`() {
        val neutral = buildTasteProfile(emptyList())
        val reason = buildReason(neutral, neutral, candidate("x"))
        assertEquals("A pick that fits both your tastes", reason)
    }

    // ---------- rankCandidates ----------

    @Test
    fun `rankCandidates sorts by finalScore descending`() {
        val fans = buildTasteProfile(
            listOf(
                rated("1", 90.0, genres = listOf("Sci-Fi")),
                rated("2", 88.0, genres = listOf("Sci-Fi"), releaseYear = 2022),
            ),
        )
        val result = rankCandidates(
            fans,
            fans,
            listOf(
                candidate("weak", genres = listOf("Musical"), tmdbRating = 4.0),
                candidate("strong", genres = listOf("Sci-Fi"), tmdbRating = 8.0),
                candidate("middling", genres = listOf("Sci-Fi"), tmdbRating = 5.0),
            ),
        )

        assertEquals(listOf("strong", "middling", "weak"), result.ranked.map { it.film.filmId })
        assertFalse(result.noMatches)
    }

    @Test
    fun `rankCandidates reports noMatches instead of suggesting a weak film`() {
        val mismatched = buildTasteProfile(
            listOf(
                rated("1", 5.0, genres = listOf("Musical")),
                rated("2", 2.0, genres = listOf("Musical"), releaseYear = 2022),
            ),
        )
        val result = rankCandidates(mismatched, mismatched, listOf(candidate("bad", genres = listOf("Musical"), tmdbRating = 2.0)))

        assertTrue(result.noMatches)
        assertTrue(result.ranked[0].finalScore < AlgorithmConfig.NO_MATCH_THRESHOLD)
    }

    @Test
    fun `rankCandidates reports noMatches for an empty candidate pool rather than throwing`() {
        val fans = buildTasteProfile(listOf(rated("1", 90.0, genres = listOf("Sci-Fi"))))
        val result = rankCandidates(fans, fans, emptyList())
        assertTrue(result.noMatches)
        assertTrue(result.ranked.isEmpty())
    }

    // ---------- advanceStreak ----------

    private fun at(iso: String): Instant = Instant.parse(iso)

    @Test
    fun `advanceStreak starts at 1 on the pair's first watch`() {
        val result = advanceStreak(StreakState(count = 0, lastWatchAt = null), at("2026-09-03T20:00:00Z"))
        assertEquals(StreakResult(count = 1, changed = true, reason = "first_watch"), result)
    }

    @Test
    fun `advanceStreak continues on the next day`() {
        val state = StreakState(count = 4, lastWatchAt = at("2026-09-02T20:00:00Z"))
        val result = advanceStreak(state, at("2026-09-03T20:00:00Z"))
        assertEquals(5, result.count)
        assertEquals("continued", result.reason)
    }

    @Test
    fun `advanceStreak counts calendar days, not elapsed hours`() {
        // 23:00 on the 2nd to 20:00 on the 3rd is only 21 hours, but two days.
        val state = StreakState(count = 2, lastWatchAt = at("2026-09-02T23:00:00Z"))
        val result = advanceStreak(state, at("2026-09-03T20:00:00Z"))
        assertEquals(3, result.count)
        assertEquals("continued", result.reason)
    }

    @Test
    fun `advanceStreak does not count a second film on the same day`() {
        val state = StreakState(count = 3, lastWatchAt = at("2026-09-03T19:00:00Z"))
        val result = advanceStreak(state, at("2026-09-03T22:30:00Z"))
        assertEquals(StreakResult(count = 3, changed = false, reason = "same_day"), result)
    }

    @Test
    fun `advanceStreak survives a gap inside the grace window`() {
        val state = StreakState(count = 6, lastWatchAt = at("2026-09-03T20:00:00Z"))
        val result = advanceStreak(state, at("2026-09-08T20:00:00Z"))
        assertEquals(7, result.count)
        assertEquals("continued", result.reason)
    }

    @Test
    fun `advanceStreak resets once the gap exceeds the grace window`() {
        val state = StreakState(count = 12, lastWatchAt = at("2026-09-03T20:00:00Z"))
        val result = advanceStreak(state, at("2026-09-14T20:00:00Z"))
        assertEquals(StreakResult(count = 1, changed = true, reason = "reset"), result)
    }

    @Test
    fun `advanceStreak treats exactly the grace window as still alive`() {
        val state = StreakState(count = 2, lastWatchAt = at("2026-09-03T20:00:00Z"))
        assertEquals("continued", advanceStreak(state, at("2026-09-10T20:00:00Z")).reason)
        assertEquals("reset", advanceStreak(state, at("2026-09-11T20:00:00Z")).reason)
    }

    @Test
    fun `advanceStreak resets rather than growing if the clock moves backwards`() {
        val state = StreakState(count = 5, lastWatchAt = at("2026-09-10T20:00:00Z"))
        val result = advanceStreak(state, at("2026-09-03T20:00:00Z"))
        assertEquals(StreakResult(count = 1, changed = true, reason = "reset"), result)
    }

    @Test
    fun `advanceStreak honours a stricter grace window when configured`() {
        val state = StreakState(count = 3, lastWatchAt = at("2026-09-03T20:00:00Z"))
        assertEquals("continued", advanceStreak(state, at("2026-09-04T20:00:00Z"), streakGraceDays = 1).reason)
        assertEquals("reset", advanceStreak(state, at("2026-09-05T20:00:00Z"), streakGraceDays = 1).reason)
    }

    @Test
    fun `advanceStreak pins the documented deviation - UTC calendar days, no per-pair timezone`() {
        // The server's timezone-aware equivalent returns "continued" for a
        // Tokyo-timezoned pair (already the next day there) for this exact
        // instant, but "same_day" for UTC. This port has no timezone
        // parameter, so it always behaves like the UTC case.
        val state = StreakState(count = 1, lastWatchAt = at("2026-09-03T14:00:00Z"))
        val result = advanceStreak(state, at("2026-09-03T23:30:00Z"))
        assertEquals("same_day", result.reason)
    }
}

File: web/match-engine.test.js
import { describe, expect, it } from "vitest";
import {
  ALGORITHM_CONFIG,
  advanceStreak,
  buildReason,
  buildTasteProfile,
  decadeKey,
  predictScore,
  rankCandidates,
  scoreCandidate,
} from "./match-engine.js";

/**
 * Ported from functions/test/{tasteProfile,scoring,streak}.test.ts — the
 * server's own tests for the exact logic this file re-implements. Adapted
 * for the two documented deviations: no country signal (so no countries
 * field, no country-weighting assertions), and advanceStreak takes no
 * timezone parameter (keys off UTC calendar days always — see the last
 * test in the advanceStreak block, which pins that deviation explicitly).
 */

const film = (over) => ({
  filmId: over.filmId,
  genres: ["Drama"],
  releaseYear: 2021,
  tmdbRating: 7,
  ...over,
});

const candidate = (over) => ({
  filmId: over.filmId,
  title: over.title || over.filmId,
  posterPath: null,
  genres: ["Sci-Fi"],
  releaseYear: 2021,
  tmdbRating: 7,
  ...over,
});

describe("decadeKey", () => {
  it("buckets modern years by decade", () => {
    expect(decadeKey(2024)).toBe("2020s");
    expect(decadeKey(2020)).toBe("2020s");
    expect(decadeKey(2019)).toBe("2010s");
    expect(decadeKey(2000)).toBe("2000s");
  });

  it("collapses everything before 2000 into one bucket", () => {
    expect(decadeKey(1999)).toBe("pre-2000");
    expect(decadeKey(1954)).toBe("pre-2000");
  });

  it("does not throw on garbage years", () => {
    expect(decadeKey(Number.NaN)).toBe("pre-2000");
  });
});

describe("buildTasteProfile", () => {
  it("averages scores per genre across films", () => {
    const profile = buildTasteProfile([
      film({ filmId: "1", score: 80, genres: ["Sci-Fi"] }),
      film({ filmId: "2", score: 60, genres: ["Sci-Fi"] }),
      film({ filmId: "3", score: 20, genres: ["Comedy"] }),
    ]);

    expect(profile.genreAffinity["Sci-Fi"]).toBe(70);
    expect(profile.genreAffinity["Comedy"]).toBe(20);
    expect(profile.sampleSize).toBe(3);
  });

  it("credits a multi-genre film to every one of its genres", () => {
    const profile = buildTasteProfile([
      film({ filmId: "1", score: 90, genres: ["Sci-Fi", "Drama"] }),
    ]);

    expect(profile.genreAffinity["Sci-Fi"]).toBe(90);
    expect(profile.genreAffinity["Drama"]).toBe(90);
  });

  it("builds era affinity alongside genre", () => {
    const profile = buildTasteProfile([
      film({ filmId: "1", score: 85, releaseYear: 2022 }),
      film({ filmId: "2", score: 45, releaseYear: 1994 }),
    ]);

    expect(profile.eraAffinity["2020s"]).toBe(85);
    expect(profile.eraAffinity["pre-2000"]).toBe(45);
  });

  it("returns an empty profile for a user who has rated nothing", () => {
    const profile = buildTasteProfile([]);
    expect(profile.genreAffinity).toEqual({});
    expect(profile.sampleSize).toBe(0);
  });
});

describe("predictScore", () => {
  const target = candidate({ filmId: "c1" });

  it("applies the configured genre/era signal weights", () => {
    const profile = buildTasteProfile([
      film({ filmId: "1", score: 100, genres: ["Sci-Fi"], releaseYear: 2021 }),
    ]);
    // Both dimensions are 100, so the weighted sum is 100 regardless of split.
    expect(predictScore(profile, target)).toBeCloseTo(100, 6);
  });

  it("weights genre more heavily than era", () => {
    const genreOnly = buildTasteProfile([
      film({ filmId: "1", score: 100, genres: ["Sci-Fi"], releaseYear: 1980 }),
    ]);
    const eraOnly = buildTasteProfile([
      film({ filmId: "1", score: 100, genres: ["Horror"], releaseYear: 2021 }),
    ]);

    expect(predictScore(genreOnly, target)).toBeGreaterThan(predictScore(eraOnly, target));
  });

  it("falls back to the neutral midpoint for unseen dimensions", () => {
    const profile = buildTasteProfile([]);
    expect(predictScore(profile, target)).toBeCloseTo(ALGORITHM_CONFIG.neutralAffinity, 6);
  });

  it("stays inside 0-100 for the extremes of the Taste Dial", () => {
    const hated = buildTasteProfile([
      film({ filmId: "1", score: 0, genres: ["Sci-Fi"], releaseYear: 2021 }),
    ]);
    const loved = buildTasteProfile([
      film({ filmId: "1", score: 100, genres: ["Sci-Fi"], releaseYear: 2021 }),
    ]);

    expect(predictScore(hated, target)).toBeGreaterThanOrEqual(0);
    expect(predictScore(loved, target)).toBeLessThanOrEqual(100);
  });
});

describe("scoreCandidate", () => {
  it("penalises divergence: a shared 70/70 beats a lopsided 95/45", () => {
    const sciFiFan = buildTasteProfile([
      film({ filmId: "a1", score: 95, genres: ["Sci-Fi"] }),
      film({ filmId: "a2", score: 95, genres: ["Sci-Fi"], releaseYear: 2022 }),
    ]);
    const sciFiSkeptic = buildTasteProfile([
      film({ filmId: "b1", score: 45, genres: ["Sci-Fi"] }),
      film({ filmId: "b2", score: 45, genres: ["Sci-Fi"], releaseYear: 2022 }),
    ]);
    const bothLukewarm = buildTasteProfile([
      film({ filmId: "c1", score: 70, genres: ["Sci-Fi"] }),
      film({ filmId: "c2", score: 70, genres: ["Sci-Fi"], releaseYear: 2022 }),
    ]);

    const target = candidate({ filmId: "x" });
    const lopsided = scoreCandidate(sciFiFan, sciFiSkeptic, target);
    const shared = scoreCandidate(bothLukewarm, bothLukewarm, target);

    // Both average to the same predicted score, but only one is a shared taste.
    expect(lopsided.breakdown.predictedA + lopsided.breakdown.predictedB).toBeCloseTo(
      shared.breakdown.predictedA + shared.breakdown.predictedB,
      6
    );
    expect(shared.finalScore).toBeGreaterThan(lopsided.finalScore);
  });

  it("applies the divergence penalty at the configured rate", () => {
    const a = buildTasteProfile([film({ filmId: "1", score: 100, genres: ["Sci-Fi"] })]);
    const b = buildTasteProfile([film({ filmId: "1", score: 0, genres: ["Sci-Fi"] })]);
    const result = scoreCandidate(a, b, candidate({ filmId: "x" }));

    const { predictedA, predictedB, divergence, tasteScore } = result.breakdown;
    expect(divergence).toBeCloseTo(Math.abs(predictedA - predictedB), 6);
    expect(tasteScore).toBeCloseTo(
      (predictedA + predictedB) / 2 - divergence * ALGORITHM_CONFIG.divergencePenalty,
      6
    );
  });

  it("scales the TMDB quality bonus onto the same 0-100 axis as taste", () => {
    const profile = buildTasteProfile([film({ filmId: "1", score: 70, genres: ["Sci-Fi"] })]);
    const perfect = scoreCandidate(profile, profile, candidate({ filmId: "c1", tmdbRating: 10 }));
    const awful = scoreCandidate(profile, profile, candidate({ filmId: "c2", tmdbRating: 0 }));

    expect(perfect.breakdown.qualityBonus).toBe(100);
    expect(awful.breakdown.qualityBonus).toBe(0);
    expect(perfect.finalScore - awful.finalScore).toBeCloseTo(100 * ALGORITHM_CONFIG.qualityWeight, 6);
  });

  it("lets shared taste outrank general acclaim, as the low quality weight intends", () => {
    const loversOfHorror = buildTasteProfile([
      film({ filmId: "1", score: 95, genres: ["Horror"] }),
      film({ filmId: "2", score: 95, genres: ["Horror"], releaseYear: 2022 }),
    ]);
    const belovedButWrongGenre = candidate({ filmId: "acclaimed", genres: ["Musical"], tmdbRating: 9.5 });
    const ourKindOfFilm = candidate({ filmId: "ours", genres: ["Horror"], tmdbRating: 6.0 });

    const acclaimed = scoreCandidate(loversOfHorror, loversOfHorror, belovedButWrongGenre);
    const ours = scoreCandidate(loversOfHorror, loversOfHorror, ourKindOfFilm);

    expect(ours.finalScore).toBeGreaterThan(acclaimed.finalScore);
  });

  it("never returns a score outside 0-100", () => {
    const hater = buildTasteProfile([film({ filmId: "1", score: 0, genres: ["Sci-Fi"] })]);
    const lover = buildTasteProfile([film({ filmId: "1", score: 100, genres: ["Sci-Fi"] })]);

    for (const [a, b] of [
      [hater, hater],
      [lover, lover],
      [hater, lover],
    ]) {
      for (const tmdbRating of [0, 5, 10]) {
        const { finalScore } = scoreCandidate(a, b, candidate({ filmId: "x", tmdbRating }));
        expect(finalScore).toBeGreaterThanOrEqual(0);
        expect(finalScore).toBeLessThanOrEqual(100);
      }
    }
  });
});

describe("buildReason", () => {
  it("names the shared genre when affinity clears the strong threshold", () => {
    const fans = buildTasteProfile([
      film({ filmId: "1", score: 90, genres: ["Comedy"] }),
      film({ filmId: "2", score: 90, genres: ["Comedy"], releaseYear: 2022 }),
    ]);
    const reason = buildReason(fans, fans, candidate({ filmId: "x", genres: ["Comedy"] }));
    expect(reason).toBe("You both love Comedy");
  });

  it("falls back to era when no genre clears the threshold", () => {
    const fans = buildTasteProfile([
      film({ filmId: "1", score: 90, genres: ["Drama"], releaseYear: 1985 }),
      film({ filmId: "2", score: 90, genres: ["Drama"], releaseYear: 1988 }),
    ]);
    const reason = buildReason(
      fans,
      fans,
      candidate({ filmId: "x", genres: ["Western"], releaseYear: 1987 })
    );
    expect(reason).toBe("You both gravitate toward pre-2000s films");
  });

  it("gives a generic reason when nothing clears either threshold", () => {
    const neutral = buildTasteProfile([]);
    const reason = buildReason(neutral, neutral, candidate({ filmId: "x" }));
    expect(reason).toBe("A pick that fits both your tastes");
  });
});

describe("rankCandidates", () => {
  const fans = buildTasteProfile([
    film({ filmId: "1", score: 90, genres: ["Sci-Fi"] }),
    film({ filmId: "2", score: 88, genres: ["Sci-Fi"], releaseYear: 2022 }),
  ]);

  it("sorts by finalScore descending", () => {
    const result = rankCandidates(fans, fans, [
      candidate({ filmId: "weak", genres: ["Musical"], tmdbRating: 4 }),
      candidate({ filmId: "strong", genres: ["Sci-Fi"], tmdbRating: 8 }),
      candidate({ filmId: "middling", genres: ["Sci-Fi"], tmdbRating: 5 }),
    ]);

    expect(result.ranked.map((c) => c.film.filmId)).toEqual(["strong", "middling", "weak"]);
    expect(result.noMatches).toBe(false);
  });

  it("reports noMatches instead of suggesting a weak film", () => {
    const mismatched = buildTasteProfile([
      film({ filmId: "1", score: 5, genres: ["Musical"] }),
      film({ filmId: "2", score: 2, genres: ["Musical"], releaseYear: 2022 }),
    ]);
    const result = rankCandidates(mismatched, mismatched, [
      candidate({ filmId: "bad", genres: ["Musical"], tmdbRating: 2 }),
    ]);

    expect(result.noMatches).toBe(true);
    expect(result.ranked[0].finalScore).toBeLessThan(ALGORITHM_CONFIG.noMatchThreshold);
  });

  it("reports noMatches for an empty candidate pool rather than throwing", () => {
    const result = rankCandidates(fans, fans, []);
    expect(result.noMatches).toBe(true);
    expect(result.ranked).toHaveLength(0);
  });
});

describe("advanceStreak", () => {
  const at = (iso) => new Date(iso);

  it("starts at 1 on the pair's first watch", () => {
    const result = advanceStreak({ count: 0, lastWatchAt: null }, at("2026-09-03T20:00:00Z"));
    expect(result).toEqual({ count: 1, changed: true, reason: "first_watch" });
  });

  it("continues on the next day", () => {
    const state = { count: 4, lastWatchAt: at("2026-09-02T20:00:00Z") };
    const result = advanceStreak(state, at("2026-09-03T20:00:00Z"));
    expect(result.count).toBe(5);
    expect(result.reason).toBe("continued");
  });

  it("counts calendar days, not elapsed hours", () => {
    // 23:00 on the 2nd to 20:00 on the 3rd is only 21 hours, but two days.
    const state = { count: 2, lastWatchAt: at("2026-09-02T23:00:00Z") };
    const result = advanceStreak(state, at("2026-09-03T20:00:00Z"));
    expect(result.count).toBe(3);
    expect(result.reason).toBe("continued");
  });

  it("does not count a second film on the same day", () => {
    const state = { count: 3, lastWatchAt: at("2026-09-03T19:00:00Z") };
    const result = advanceStreak(state, at("2026-09-03T22:30:00Z"));
    expect(result).toEqual({ count: 3, changed: false, reason: "same_day" });
  });

  it("survives a gap inside the grace window", () => {
    const state = { count: 6, lastWatchAt: at("2026-09-03T20:00:00Z") };
    const result = advanceStreak(state, at("2026-09-08T20:00:00Z"));
    expect(result.count).toBe(7);
    expect(result.reason).toBe("continued");
  });

  it("resets once the gap exceeds the grace window", () => {
    const state = { count: 12, lastWatchAt: at("2026-09-03T20:00:00Z") };
    const result = advanceStreak(state, at("2026-09-14T20:00:00Z"));
    expect(result).toEqual({ count: 1, changed: true, reason: "reset" });
  });

  it("treats exactly the grace window as still alive", () => {
    const state = { count: 2, lastWatchAt: at("2026-09-03T20:00:00Z") };
    expect(advanceStreak(state, at("2026-09-10T20:00:00Z")).reason).toBe("continued");
    expect(advanceStreak(state, at("2026-09-11T20:00:00Z")).reason).toBe("reset");
  });

  it("resets rather than growing if the clock moves backwards", () => {
    const state = { count: 5, lastWatchAt: at("2026-09-10T20:00:00Z") };
    const result = advanceStreak(state, at("2026-09-03T20:00:00Z"));
    expect(result).toEqual({ count: 1, changed: true, reason: "reset" });
  });

  it("honours a stricter grace window when configured", () => {
    const strict = { streakGraceDays: 1 };
    const state = { count: 3, lastWatchAt: at("2026-09-03T20:00:00Z") };
    expect(advanceStreak(state, at("2026-09-04T20:00:00Z"), strict).reason).toBe("continued");
    expect(advanceStreak(state, at("2026-09-05T20:00:00Z"), strict).reason).toBe("reset");
  });

  it("pins the documented deviation: UTC calendar days, no per-pair timezone", () => {
    // streak.test.ts's server equivalent of this exact instant returns
    // "continued" for a Tokyo-timezoned pair (already the next day there)
    // but "same_day" for UTC. The web version has no timezone parameter, so
    // it always behaves like the UTC case, even for a pair that isn't in UTC.
    const state = { count: 1, lastWatchAt: at("2026-09-03T14:00:00Z") };
    const result = advanceStreak(state, at("2026-09-03T23:30:00Z"));
    expect(result.reason).toBe("same_day");
  });
});
